File: Models/Hotel.cs

using System.ComponentModel.DataAnnotations;
using System.Text.Json.Serialization;
using Microsoft.AspNetCore.Mvc.ModelBinding;

namespace RoomMonitor.Models
{

    // interface IBaseHotelKey {
    //     string HotelChain { get; set; }
    //     string CountryCode { get; set; }]
    //     string Town { get; set; }
    //     string Suburb { get; set; }
    // }
    public abstract class BaseHotelKey
    {

        [StringLength(100)]
        public abstract string HotelChain { get; set; }

        [StringLength(2)]
        public abstract string CountryCode { get; set; }

        [StringLength(100)]
        public abstract string Town { get; set; }

        [StringLength(100)]
        public abstract string Suburb { get; set; }
    }

    public class HotelKey : BaseHotelKey
    {
        [BindRequired]
        public override string HotelChain { get; set; }

        [BindRequired]
        public override string CountryCode { get; set; }

        [BindRequired]
        public override string Town { get; set; }

        [BindRequired]
        public override string Suburb { get; set; }
    }

    public class Hotel : BaseHotelKey
    {
        [Required]
        public override string HotelChain { get; set; }

        [Required]
        public override string CountryCode { get; set; }

        [Required]
        public override string Town { get; set; }

        [Required]
        public override string Suburb { get; set; }

        [Required]
        [Range(0, 5)]
        public int? NumStar { get; set; }

        public HotelKey GetKey() => new HotelKey
        {
            HotelChain = this.HotelChain,
            CountryCode = this.CountryCode,
            Town = this.Town,
            Suburb = this.Suburb
        };
    }
}

File: ClientApp/src/app/shared/models/Hotel.ts
import { ICompositeKey } from "./../interfaces/ICompositeKey";
import { IDeserializable } from "../interfaces/IDeserializable";

export interface IHotelKey {
  hotelChain: string;
  countryCode: string;
  town: string;
  suburb: string;
}

export interface IHotel extends IHotelKey {
  numStar: number;
}

export class Hotel
  implements IHotel, IDeserializable<IHotel>, ICompositeKey<IHotelKey> {
  hotelChain: string;
  countryCode: string;
  town: string;
  suburb: string;
  numStar: number;

  deserialize(input: IHotel): this {
    return Object.assign(this, input);
  }

  getKey() {
    return {
      hotelChain: this.hotelChain,
      countryCode: this.countryCode,
      town: this.town,
      suburb: this.suburb,
    };
  }
}
